### ISSUE TITLE
[Verify/Document] Native (conditional) dependency support. 

### ISSUE THREAD

**Issue by nathanaeljones** [NuGet/Home#300]
Originally [posted on Paket](https://github.com/fsprojects/Paket/issues/714), but since this is something NuGet is considering as well, I'm adding it here so we can track and (hopefully sync efforts). I hear that the ASP.NET team is doing something with native  dependencies, but I can't find a spec for it. I suspect that a lockfile and transitive dependency support are pre-requisites for this. 

A lot has been written about this (challenging) topic; so I'll link to what I've found (and please add more links in the comments).
- https://stackoverflow.com/questions/19478775/add-native-files-from-nuget-package-to-project-output-directory
- https://github.com/aspnet/dnx/issues/402
- https://nuget.codeplex.com/workitem/1221
- https://nuget.codeplex.com/workitem/679
- https://nuget.codeplex.com/discussions/412012
- http://ilnumerics.net/blog/anycpu-computing-limping-platform-specific-targets-and-a-happy-deployment-end/
- http://youku.io/questions/392214/nuget-powershell-how-to-add-native-dependencies-how-to-add-files-to-a-project
- https://nuget.codeplex.com/discussions/446656
- https://github.com/MvvmCross/MvvmCross/issues/307
- https://nuget.codeplex.com/discussions/286521
- https://nuget.codeplex.com/discussions/230521
- https://nuget.codeplex.com/workitem/109

I've been using a https download-during-boot approach for ImageResizer, but that is slow, unreliable, and annoying. I tried to [create an example of how to build the ideal native/managed hybrid project](https://github.com/imazen/paragon), failed, then [started a project to try to hot-fix the problem at runtime](https://github.com/imazen/Imazen.NativeDependencyManager), and [hit another series of roadblocks](https://github.com/imazen/Imazen.NativeDependencyManager/issues/5). 

I've identified a few invalid assumptions that seem responsible for the current state of things. 

Some of these are somewhat comical considering [how easy it is to parse binaries for the major platforms and determine runtime compatibility](https://github.com/imazen/Imazen.NativeDependencyManager). 
1. It is bad to assume all _managed_ dlls are AnyCPU. x86 and x64-only binaries continue to have an important place. a) C++/CLI is still important, but only targets Win32 and x64. b) [Binding generation tools like CppSharp](https://github.com/mono/CppSharp)  cannot produce AnyCPU C# - the structure layouts (and other details) are calculated based on a pointer width assumption.  c) There's also manually written C# that uses unsafe code or performs P/Invokes, and doesn't use IntPtr in all the right places, and is therefore x86 or x64-only.
2. It is bad to assume that precompiled native binaries are appropriate for every operating system distribution. Source compilation needs to be first-class if we want to target a large number of linux or bsd distributions; we can't precompile for everything.
3. It is bad to assume that binaries are small enough that all the permutations can go in the same zipped nuget package.  OpenCV base libraries could easily be > 2 gigabytes if you did this. You'd also exhaust your server's disk space. And likely your bandwidth allotment. And requests would time out.
4. It is bad to assume that a windows machine is capable of compiling native code.

---

Removing these assumptions, what new requirements are we left with? 
1. Nuget packages need to be able to reference other nuget packages - conditionally - based on architecture and operating system. This likely means that we need conditions in the lockfile, too - use version X of package Y on platform Z, etc. At build time, the right versions are copied.
2. Nuget packages need to be able to describe native binaries - or rather - arbitrary files, and provide multiple versions for each target platform.  For small binaries, a common pattern is likely to combine these two approaches, and provide x86/x64 binaries for windows, and a conditional reference for other platforms. For larger binaries, it's likely that the 'main' package will be empty, and simply list conditional, plat-specific packages as dependencies.
3. Build time. This is a bit harder. What needs to happen?
   a) We need to gather the referenced files (nuget references, mind you), and verify that the output folder does not have any conflicting named files. If there are conflicting names, the output folder version MUST be deleted, so that we can AssemblyResolve or LoadLibrary the correct version. We then copy each of the files to an appropriate subfolder of the output folder (or, if AnyCPU, the output folder itself).  Since VisualStudio is blind to compatibility (by choice, one must assume), we may end up fighting with the build process a bit. Perhaps disabling copylocal? Another nice sanity check would be to simply parse the binary headers of everything in the output folder and ensure they are all able to run on a common environment. 
4. Run time. This is where developers have to do a little magic, and call AssemblyResolve for non-AnyCPU managed dlls, and LoadLibrary (or a platform alternative) for the native dependencies. Given a standardized convention, we can make this co-operation possible without .NET runtime changes (although not optimal, since Assembly.Load won't use the default security context). It would be much better if .NET would modify the search path based on platform, as it does for culture. It would also be great if ASP.NET would apply some intelligence or header parsing to assemblies before globbing them all into memory. It's not hard, and only requires between 1 and 3 I/O reads of a few hundred bytes.
5. Tooling. Tooling needs to understand that there are non-.NET dependencies involved. Unit test runners, in particular, are known for leaving behind native dependencies when they copy (or shadow copy) assembles for testing. These will need to understand the (transitive) dependencies. Perhaps we should emit some kind of manifest? While we can piggy-back on .NET assemblies to document dependencies (via resource manifests or regular assembly attributes), those .NET assemblies would need to document the final transitive set of native dependencies, which might be difficult to achieve. 

So, I guess 
- Step 1: Establish (a) plat/arch conditions, and (b) target strings we can use as subfolders.
- Step 2: Describe required changes to the nuget package specification
- Step 3: Implement handling in paket for nuget conditions and native file manifests, through to lockfile.
- Step 4: Implement paket support for triggering bash/.bat build scripts (perhaps requiring user authorization on a per-commit-id basis, if applied to a git reference)? CMake is probably most likely to be used, but if we require Git, we can require Bash.
- Step 5: Implement build-time support for copying and manifest generation.
- Step 6: Submit PRs to major runners and web frameworks to handle manifests and assembly loading properly.
- Step 7: Make .NET into a real platform, where C interop is practical, so we can play with the big kids.
### Conditions:
- pointerSize=32|64. Let's say that our MSIL makes an assumption about pointer size, but not platform.
- endianess="little|big" Sometimes this matters more than architecture, and ARM can switch between endianess modes.
- architecture=x86|x64|IA64|Alpha|MIPS|HPPA|PowerPC|SPARC32|s390|s390x  - Should probably support everything [Mono does](http://www.mono-project.com/docs/about-mono/supported-platforms/)
- os="posix|winish|linux|osx|win7|win8|win10" I'm not sure how to best divide windows (or linux) operating systems into groups, or to number them for easy inequality testing, but we probably want to establish sane identifiers that correspond to common build/api compat targets.

Given that a fallback mode (building from source) is likely popular, we want to make it easy to ensure that only 1 reference from a conditional set is chosen. We should probably group them within another element or provide an id that prevents duplicates. 
### Target strings

Target string need to be as generic as can be permitted based on their restrictions. 

/  - root is AnyCPU
/32b/ - managed, assumes 32-bit pointer, otherwise portable
/64b/ - managed, assumes 64-bit pointer, otherwise portable
/x86/winish/ - 32-bit, requires windows APIs. 

pointer size, architecutre, and endianess are combined into the first string. Pointer size and endianess are only included if the architecture string doesn't make them redundant. I.e, we would see /ARM-little/ and /ARM-big/, but not /x86-little/.

**Comment by davidfowl** [comment]
@nathanaeljones Have you done any case studies of other package managers that may have these same problems? Do they do conditional dependencies based on arbitrary pivots? Does that condition on the dependency edge affect version selection? 

I know npm supports native modules being compiled on the target machine at install time. Are you suggesting that we support the full spectrum of things?

> a) We need to gather the referenced files (nuget references, mind you), and verify that the output folder does not have any conflicting named files. If there are conflicting names, the output folder version MUST be deleted, so that we can AssemblyResolve or LoadLibrary the correct version. We then copy each of the files to an appropriate subfolder of the output folder (or, if AnyCPU, the output folder itself). Since VisualStudio is blind to compatibility (by choice, one must assume), we may end up fighting with the build process a bit. Perhaps disabling copylocal? Another nice sanity check would be to simply parse the binary headers of everything in the output folder and ensure they are all able to run on a common environment.

We're enhancing MSBuild in vs14 with NuGet aware targets that understand runtime/vs compile time.

> Run time. This is where developers have to do a little magic, and call AssemblyResolve for non-AnyCPU managed dlls, and LoadLibrary (or a platform alternative) for the native dependencies. Given a standardized convention, we can make this co-operation possible without .NET runtime changes (although not optimal, since Assembly.Load won't use the default security context). It would be much better if .NET would modify the search path based on platform, as it does for culture. It would also be great if ASP.NET would apply some intelligence or header parsing to assemblies before globbing them all into memory. It's not hard, and only requires between 1 and 3 I/O reads of a few hundred bytes.

This is what we do today with things like kestrel in ASP.NET 5. See my comment here https://github.com/imazen/Imazen.NativeDependencyManager/issues/5#issuecomment-87291776. I'm not sure about making these new things work down level on all CLRs but newer ones might be more viable.

> Tooling. Tooling needs to understand that there are non-.NET dependencies involved. Unit test runners, in particular, are known for leaving behind native dependencies when they copy (or shadow copy) assembles for testing. These will need to understand the (transitive) dependencies. Perhaps we should emit some kind of manifest? While we can piggy-back on .NET assemblies to document dependencies (via resource manifests or regular assembly attributes), those .NET assemblies would need to document the final transitive set of native dependencies, which might be difficult to achieve.

Transitive dependencies are going to be understood with build time and runtime nuget resolution coming into the platform.

I'm actually not a fan of complicating the graph walk by adding arbitrary pivots to it. I think that's actually an intractable problem. I do however agree with the fact that we do need conventions to describe these things in NuGet packages and have appropriate behaviors at runtime and build time.

**Comment by nathanaeljones** [comment]
Conditional packages based on platform seems to be very widespread. [npm has optionaldependencies and platform filtering](http://stackoverflow.com/questions/15176082/npm-package-json-os-specific-dependency), [bundler has :platform=> ](http://stackoverflow.com/questions/8421321/does-using-platforms-in-your-gemfile-work). 

The problem with Bundler, (and the reason windows support is troubled) is its lack of [support for pivots *within* the lockfile](https://github.com/bundler/bundler/issues/635). Bundler supports platform pivots, but doesn't generate different lockfiles for each target. I think this is considered a flaw by the Bundler team (although I can't find the link; could have been mentioned at rubyconf 2014).

There are the choices I see:
1. Build from source.
2. Don't do anything operating-system specific, never use native code.
3. Have pivots. 
4. Have both pivots and permit building from source.

I'm a fan of #4. When it's small, building from source is fine. But msvc can be 10-50x slower than gcc; building a tiny graphics library (libgd) with 3 dependencies takes over 45 minutes on windows. On linux, that's 4 minutes. Both options should be present.

I'm also not suggesting arbitrary pivots; rather just 4 pivots. That's a manageable number, I think. If you limit it to modern platforms, there are only 12 permutations. In practice, we're going to see binaries for windows, and source code for everything else. That's 3 permutations, 4-5 if you include ARM processors. I'd expect binaries for *nix platforms only where compilation is excessively slow. 

Is the graph walk so resource-intensive that we can't take a 3-5x hit for the portions of that graphic which have conditional-based native dependencies? 

I'm not sure I see how the graph is an intractable problem, especially if we only have to "solve" it for the targets that the consuming project defines. Hey, &mdash; if the nuget server can send back dependency information for a package without sending the whole package &mdash; even  the most naive approach should be just fine.

You should really talk to @wycats. He's the library package manager design expert, and he's dealt with this problem in many forms.

**Comment by davidfowl** [comment]
Walking the graph isn't hard, picking versions based on compatible pivots is hard and leads to unpredictability in the algorithm, which I think is much worse.

**Comment by nathanaeljones** [comment]
I would not select versions based on any of these pivots. I'm not even fully convinced that picking versions based on the .NET framework version is a good idea.

**Comment by nathanaeljones** [comment]
It's far better to "let it fail" than to make the algorithm unpredictable. If it fails, it can be fixed. If nothing fails, there is no pain point to repair, and that platform may remain stuck in the past forever.

**Comment by yishaigalatzer** [comment]
We are going to do something for RTM for sure. Will loop back when there is more meaty code to talk about.

**Comment by nathanaeljones** [comment]
I hope this can be designed in the open.

**Comment by nathanaeljones** [comment]
Are there any specs/documents/drafts we can review?

**Comment by yishaigalatzer** [comment]
Not yet, still working on it

**Comment by yishaigalatzer** [comment]
@nathanaeljones we are going to have project.json and a lock file as well as transitive restore supporting install.ps1 uninstall.ps1 content and targets fully in Visual Studio. The code is already public in NuGet but we are still playing around with it. It requires an msbuild task/target file that is being worked on as well (but is not public yet).

I don't understand the native picture 100% so I don't want to add misleading information until its properly documented.

**Comment by nathanaeljones** [comment]
@yishaigalatzer Have you resolved the issues with the msbuild team? Last I heard they were refusing to support NuGet (or any package manager, for that matter). I think the requirement was to enable reloading of the build file once the package manager has finished its work. [Packet and NuGet both ran into the same issue, I think](https://github.com/fsprojects/Paket/issues/417#issuecomment-66473466). 

We'd all like to see a solution that can work without a VS-specific extension, but right now it appears to be at an impasse.

----

The other issue is that [AppDomainSetup](https://msdn.microsoft.com/en-us/library/system.appdomainsetup%28v=vs.110%29.aspx) ShadowCopy functionality needs to be aware of that transitive list.  [xUnit's usage here](https://github.com/xunit/xunit/blob/9b090186792039db5699c0a418591eb566f23c36/src/common/RemoteAppDomainManager.cs#L37-L50). We've [exhaustively explored workarounds to hack this in lieu of a .NET 4.6 feature addition](https://github.com/imazen/Imazen.NativeDependencyManager/issues/5#issuecomment-87291776), and the smallest change we can come up with is to add a read-only field to Assembly that provides the original (pre-shadow-copy) path of the assembly. Without that, there's no way to guess where the assembly came from, and where the missing files could be located.

Ideally, of course, the runtime would be aware of this lockfile and handle the xplat differences for us. 

---- 

More detail about install/uninstall would be great. First-class [native binary nuget packages](https://github.com/dotnet/corefx/issues/1298) would be helpful from a security point of view; wrappers wouldn't need to update for a security patch to be applied.

**Comment by yishaigalatzer** [comment]
There is no msbuild involvement. We run nuget restore before build in nuget 3. Try it out in the rc bits.

I'll read about the other issues you mentioned.

Paket is honestly a very limited view on what NuGet packages contain. And I can't speak to what its limitations are.

I'll make sure that once command line bits are available we will reach out to you with more details.

**Comment by yishaigalatzer** [comment]
We are now building NuGet based on lock files and project.json (as you can see in all our client repos). The release is going to be a bit after RTM, but the bits are already somewhat working.

We are not currently supporting pivots. But we are starting to finally working on documenting the process of what we got so far.

**Comment by yishaigalatzer** [comment]
So I don't think this is going to get fully resolved in the first release. I'm still leaving it at that milestone so we can keep it in the front, and turn this issue in a feedback/next steps kind of issues on top of that release

**Comment by yishaigalatzer** [comment]
@nathanaeljones here is a link to a package that addresses the native scenario -
https://www.myget.org/F/dotnet-core/api/v2/package/System.IO.Compression/4.0.0-beta-22927

**Comment by ghost** [comment]
:+1: for having npm inspired `optionalDependency`, which do not get installed when certain CLI flag is not present during the installation. The program is supposedly work fine without optional dependencies. Think of it as MEF architecture, where if the binary is not present, some feature (say a menu item) will be absent, but the program as a whole will still function.

But

> npm has optional dependencies and platform filtering

npm does not select native binary based on the os+architecture. Do you think it creates a package dynamically everytime you run npm install and make decisions which binary to pack? No. All packages using node-gyp for C++ binary dependency have this issue. From the link you provided, optionalDepenedency is something misleading in this regard. The os key is a flat filter, it just rejects the entire installation on non-permitted OS.
Last time I checked, this is one of the major challenge for modules using C++ dependencies, as the 'package' itself is a dumb / distilled form of your product and is essentially environment-agnostic. 

Then how do they know which binary to keep?
- Compile on install (require python, C++ compiler, node-gyp). Not a very successful option for Windows users.
- Download the binary during the installation using custom install script
- Package all binaries within the package (some clean up after installation <- just being pendantic). 

The decision whether the binary is compatible maybe complicated than just matching the architecture. node.js depends on v8, which has its own versions and then there are module versions etc. (`node -p process.versions` there are 5-6 constituents and a slight version change can and does break the native dependencies. There is no universal way of distributing native binary in node.js world and it is still a BIG challenge). 

-> Spoken from series of personal painful experiences with npm/C++/node-gyp.

**Comment by nathanaeljones** [comment]
@jasonwilliams200OK I assumed that you'd have a parent package with optional dependencies to every platform-specific child package. Those not relevant fail to install. Is this not possible with npm?

**Comment by jasonwilliams200OK** [comment]
@nathanaeljones, good idea; but the problem would persist that every package needs to provide its own 'sub package manager' for native binaries distribution. That sub-package-manager require building binaries for all supported "targets" which essentially consist of:

>
operatingSystems **x** architectures **x** v8-modules **x** runtime

Where runtime = node.js, io.js, AtomShell, lsc etc.

Imagine for every release, you need to generate binaries for all variants as well if you care about distributing binaries. For that reason, they have things like [node-pre-gyp](https://www.npmjs.com/package/node-pre-gyp), which mitigates the binaries building pain-point to some extent by accumulating the artifacts produces by jobs running on CI servers (TravisCI and AppVeyor).

This is a very annoying problem for people managing npm packages with C++ modules. Therefore, most of those packages require compilation at install-time for unsupported targets which npm triggers with npm commands (that is added to `install` in package.json), which require a very specific versions of python (2.6 & 2.7 coz node-gyp doesn't work with py3), gcc, clang and mvsc etc. to be installed on end-terminal.

**Comment by nathanaeljones** [comment]
@jasonwilliams200OK Yes, but as you said, only Windows desperately needs precompiled binaries. Also, why would the v8 modules matter if the native code is tangential (say a hash algorithm)? You can even have multiple C runtimes as long as you don't pass around FILE*, etc.

**Comment by jasonwilliams200OK** [comment]
@nathanaeljones, there are gazillions of breaking changes and disparities among v8 versions. For that matter we use nan https://github.com/nodejs/nan, which manage to abstract out the commonalities so C++ modules can cross compile. Going native doesn't help because we can't just tap into uv loop owned by v8 and sane-fully interact with the runtime. We have to abide by these APIs and wrappers. Once the binary is compiled, it works with only that "module version" of v8 (v8 version is different than module version which is different than nan version and so forth). Theoretically the idea is neat, practically it is not pretty. 

> only Windows desperately needs precompiled binaries

Even on Linux boxes, the consumer doesn't necessarily have python 2.6/2.7 with gcc/g++/clang (version high enough to compile code if C++11/14 features were used). It also depends on your audience. For example: node-sqlite3, node-sass, phantomjs etc. They all have different approaches of serving binaries. When consumer does `npm install node-sqlite3` on their FreeBSD running on ARM chip or SunOS on SPARC arch, they don't expect the `npm install` to fail because g++ 4.9 was missing.

**Comment by nathanaeljones** [comment]
@jasonwilliams200OK Can you invent a similar scenario that would apply to a mostly-managed .NET app? I mix and match dozens of native C libraries complied with varying versions of msvc, gcc, and even some that are cross-compiled. I haven't run into this kind of issue yet, although none of these C libraries communicate directly; they are all plugins for a .NET framework.

**Comment by lilith** [comment]
Depressing -> https://github.com/Microsoft/msbuild/issues/83

**Comment by yishaigalatzer** [comment]
No idea why you say we dropped msbuild support? We did none of that, and added a ton of features to support native directly

**Comment by lilith** [comment]
> No idea why you say we dropped msbuild support? We did none of that, and added a ton of features to support native directly

Follow the link; read jeff handley's comments. 

That wasn't intended as a criticism of NuGet - which is making great strides - but rather the continued blockade against opening the ecosystem to more specialized alternatives. 

There are dozens of things that I can only do with Paket. I'm not sure I even want NuGet to do those things, but if MSBuild makes it impossible to restore packages with an arbitrary tool, then I'm screwed. 

## I can't hand users a project that builds, and users don't try a tool twice.

Is C++/CLI _actually_ on your roadmap? Should it even be? I'm not sure at this point. The 2011 message was "wait for it, we'll do it all", followed by many elbows thrown at those where actually writing code to solve the problem. 

In 2015, maintainers have to apply linear programming to find the smallest set of forks we must maintain in parallel to reach at least 80% of the .NET user base. The several tiny, backwards-compatible changes required to let us (tediously) work around the major issues were rejected.

**Comment by yishaigalatzer** [comment]
Our suggestion is to restore before msbuild runs in a powershell script or sh script.

Why is that wrong? A user will just run build on your repo?

You can also do the same thing nuget does in visual studio (on build restore). Nuget is not getting any special accommodations from visual studio other than shipping in the box, is what you are looking for is an ability to make paket run on build even if it wasn't installed? I can think of a few alternatives on how to accomplish it.

**Comment by lilith** [comment]
> shipping in the box

^ Extremely important. 

Most of my users do not know how to open a command-line window, let alone configure a command-line environment that can build things. They double-click the Solution file and hit Run in VS. If that fails, they move on. I've watched this via screen sharing hundreds of times. 

> is what you are looking for is an ability to make paket run on build even if it wasn't installed?

Yes. If MSBuild offered a 'reload' instruction, we could bootstrap everything quite nicely. As it stands, there's no way to offer a unified experience across tooling, since MSBuild is the only common factor. 

The "Just works" factor is _the only_ factor that matters at the end of the day - particularly for the market that .NET targets. 

The Paket folks didn't write a freaking Visual Studio extension for the fun of it. They wrote it because MSBuild is petty.

**Comment by ghost** [comment]
@nathanaeljones, you might be interested in this survey: http://blogs.msdn.com/b/vcblog/archive/2015/09/18/packing-technology-survey.aspx. :)

**Comment by lilith** [comment]
Thanks. It appears to only let you pick one language, and one target platform. Particularly strange that they didn't use checkboxes given the context.

**Comment by ghost** [comment]
Eric Battalio is generally very receptive on such notes and he is reachable via email as noted here: http://blogs.msdn.com/b/vcblog/archive/2015/09/03/cross-platform-msbuild.aspx (email address provided in the post). :)

**Comment by lilith** [comment]
There are already comments to the effect on the page. Kind of a wasted survey.

**Comment by lilith** [comment]
NuGet 3 won't work on VS2013, so it's really not something we can target as library developers. 

Anything we should keep in mind or try to stay compatible with as we implement native dependency support in Paket?

**Comment by yishaigalatzer** [comment]
The story for 2013 is to use msbuild targets. If you want to target all audiences paket is not your answer. A good sample is the win2dd package.

There is nothing to do here, and this is not the paket repo. Start an issue there and we can collaborate over it.

**Comment by lilith** [comment]
Already done https://github.com/fsprojects/Paket/issues/714

Is the official NuGet client the answer to targeting all audiences? 

Anecdotally, NuGet implodes on every real-world project I give it, but Paket does just fine.  Maintaining compatibility with the default NuGet client isn't my first priority; rather, my first priority is to supply my users with a solution that 'just works' - everywhere (VS12, VS14, DNX, .NET full, .NET core, etc).

If possible, compatibility would be nice, but the existing tools are just so broken.

**Comment by yishaigalatzer** [comment]
Pushing a package that will only work on Paket is definitely a worse approach than making something that works everywhere. Yes it is harder today, and perhaps the answer is to eventually port nuget 3 down to 2013. The answer today is to use targets for down level visual studio support like win2D package does.

And as for implodes on any real-world project, I suggest filing issues, with as much detail as possible. And work through them together. The general statement above doesn't get us anywhere.

**Comment by dead-claudia** [comment]
NPM mostly delegates the step to node-gyp, which then handles OS dependencies and compiles the code itself using the system's compiler, unless a `"preinstall"` is specified in `"scripts"`, which can itself call this. Frequently, packages override this to first check for an OS-specific precompiled binary and download it if it exists. The PhantomJS runtime and node-fibers do this.

PyPI requires everything to be prebuilt, and uses setuptools to simplify and abstract the process.

I think NuGet could support something similar.

**Comment by ghost** [comment]
@impinball, like @nathanaeljones noted:

> It is bad to assume that a windows machine is capable of compiling native code.

Here is a shining example: https://github.com/nodejs/node-gyp/issues/629 with a [glimpse of hope](https://github.com/nodejs/node-gyp/issues/629#issuecomment-153196245) (in the form of Build Tools 2015 without VS).

In npm's case, it is assumed that all Unix-like operating system are packed with C/C++ compilers, while on Windows only MSBuild is supported by node-gyp. The approach you mentioned about phantomJS (and node-sass) is good that prepare the binaries for as many OSs and arch as you can and for the remaining fallback to default node-gyp's compiler from source at installation time behavior, but at the same time, too much work for the maintainer to verify and test binaries on all possible OS and architectures.

In my experience, most people are happy if they just do `npm install package` and get things rolling, preferably without compilation step. Even on Unix-likes, npm+node-gyp's advertised approach to compile on the fly fails when we use relatively new feature-set of native languages which some Linux distribution's version's out of the box compilers might not support (CentOS-5.11 which many hosting environments are running till date packs with gcc 4.4, which does not even support `nullptr`!). So every maintainer of the package (should) have to mandate the minimum C/C++ compilers' version on target as well, which does not inspire many consumers..

NuGet may experience a totally different trend of packaging stuff if this ongoing relevant discussion in CoreFX repo: https://github.com/dotnet/corefx/issues/2302#issuecomment-156132918 get popular.

**Comment by dead-claudia** [comment]
@jasonwilliams200OK Sorry to burst your bubble, but [I don't think that idea on npm is getting very many places, as great as it sounds...](https://github.com/nodejs/node-gyp/issues/629#issuecomment-157580743)

**Comment by dead-claudia** [comment]
Although I could see the benefits of .NET Core's approach.

**Comment by ghost** [comment]
@impinball, i think you totally missed the point... but this is more or less what i am trying to say..

**Comment by dead-claudia** [comment]
I'm not immediately invested in this, but this is the one thing that's stopping me from learning the platform where I use Linux (and there's a lot of .NET jobs where I live).

**Comment by lilith** [comment]
I am very glad to see [documented Runtime IDs](https://docs.microsoft.com/en-us/dotnet/articles/core/rid-catalog) have landed.

I do not see that `<group>` offers pivots by runtime ID. Is there a way to conditionally depend on a package based on runtime ID?

As discussed, it would be unexpected for the NuGet client to select a version based on what runtimes are supported by specific versions - rather, I would expect that the latest matching version number is used (based on semver evaluation), and no 'compatibility' checking is employed when selecting dependencies. If a runtime is missing for a specific version, restore fails, and the user must add a version specification.  This ensures that there is no computational complexity in flattening transitive dependencies into a lockfile despite the presence of an additional runtime pivot. 

Separating native dependencies into packages by runtime enables a very easy deployment scenario Every Travis and AppVeyor task is only responsible for uploading their own runtime's package. The last task can upload an aggregate package which references them, after testing they are all retrievable (with a retry loop).

**Comment by jainaashish** [comment]
Its been an year without any update here and not sure what was the last status since this is very long thread. So @nathanaeljones can you please check you if this still applicable for `PackageReference'? and if it does, then please open a new issue with relevant details so that we can take that forward.
